opm/simulators/wells/RateConverter.hpp

```diff
@@ -482,44 +493,63 @@ namespace Opm {
                         hydrocarbon -= fs.saturation(FluidSystem::waterPhaseIdx).value();
                     }
 
-                    int reg = cell2region[cellIdx];
+                    const auto& reg = cell2region[cellIdx];
                     assert(reg >= 0);
-                    auto& ra = attr_.attributes(reg);
-                    auto& p  = ra.pressure;
-                    auto& T  = ra.temperature;
-                    auto& rs  = ra.rs;
-                    auto& rv  = ra.rv;
-                    auto& pv  = ra.pv;
 
                     // sum p, rs, rv, and T.
-                    double hydrocarbonPV = pv_cell*hydrocarbon;
-                    if (hydrocarbonPV > 0) {
-                        pv += hydrocarbonPV;
-                        p += fs.pressure(FluidSystem::oilPhaseIdx).value()*hydrocarbonPV;
-                        rs += fs.Rs().value()*hydrocarbonPV;
-                        rv += fs.Rv().value()*hydrocarbonPV;
-                        T += fs.temperature(FluidSystem::oilPhaseIdx).value()*hydrocarbonPV;
+                    const double hydrocarbonPV = pv_cell*hydrocarbon;
+                    if (hydrocarbonPV > 0.) {
+                        auto& attr = attributes_hpv[reg];
+                        attr.pv += hydrocarbonPV;
+                        attr.p += fs.pressure(FluidSystem::oilPhaseIdx).value() * hydrocarbonPV;
+                        attr.rs += fs.Rs().value() * hydrocarbonPV;
+                        attr.rv += fs.Rv().value() * hydrocarbonPV;
+                        attr.T += fs.temperature(FluidSystem::oilPhaseIdx).value() * hydrocarbonPV;
+                    }
+
+                    if (pv_cell > 0.) {
+                        auto& attr = attributes_pv[reg];
+                        attr.pv += pv_cell;
+                        attr.p += fs.pressure(FluidSystem::oilPhaseIdx).value() * pv_cell;
```

Comment:
Any suggestion about how to handle it when the oil phase is not there? I think we did something similar in the onephase simulator case.   Should we do the same here?

Reply:
Yes we need the same logic. I think we have a consensus that it should be implemented in the FluidState though (and removed from this and other places). Could you add a method `referencePhasePressure()` to the blackoil fluid state (in opm-material), with logic to determine which should be used (that could be an independent PR)? Using it from here would be a good trial.